### ISSUE TITLE
Concatenation of "." paths possibly does not work as expected

### ISSUE THREAD

**Issue by Martinsos** [commercialhaskell/path#166]
I wanted to use Path.Windows to work with Windows paths no matter on which platform are we executing, and I also needed the same thing for Posix (Path.Posix). However, I got some unexpected (at least in my view) behaviour when using Path.Windows on Linux or Path.Posix on Windows.
Below are the tests with behaviour that I expected, and also error messages describing actual behaviour.

```hs
import qualified Path.Windows as PW
import qualified Path.Posix as PP
```

On Linux
```hs
[PW.reldir|.|] PW.</> [PW.reldir|.|] `shouldBe` [PW.reldir|.|]
```
fails with `expected ".\\" but got ".\\.\\"`. (but it passes on Windows!)

and
```hs
[PW.reldir|.|] PW.</> [PW.reldir|a|] `shouldBe` [PW.reldir|a|]
```
fails with `expected "a\\" but got ".\\a\\"`. (but it passes on Windows!)

On Windows
```hs
[PP.reldir|.|] PP.</> [PP.reldir|.|] `shouldBe` [PP.reldir|.|]
```
fails with `expected "./" but got "././"`. (but it passes on Linux!)

and
```hs
[PP.reldir|.|] PP.</> [PP.reldir|a|] `shouldBe` [PP.reldir|a|]
```
fails with `expected "a/" but got "./a/"`. (but it passes on Linux!)

**Comment by Martinsos** [comment]
One more possibly unexpected behaviour:

```hs
PW.parseRelDir "..\\a"
```
does not fail when running on Linux (it does fail on Windows), although it has .. at the beginning and it should fail.

**Comment by NorfairKing** [comment]
@Martinsos These are probably related to the failing validity-based tests. cc @mrkkrp . It would be great if we could decide on the correct behaviour here, and add these tests, before we start fixing anything.

**Comment by NorfairKing** [comment]
I'm a bit more confused now, after looking at the tests: 

https://github.com/commercialhaskell/path/blob/master/test/Posix.hs#L166-L167

The first test that you describe is right there, and it passes for posix. Am I missing something?

**Comment by Martinsos** [comment]
@NorfairKing notice that I am using functions from Path.Windows, so while test that you linked to is `($(mkRelDir "./") </> $(mkRelDir ".") == $(mkRelDir "."))`, the test that should be equivalent to one that I have failing is `($(Path.Windows.mkRelDir "./") Path.Windows.</> $(Path.Windows.mkRelDir ".") == $(Path.Windows.mkRelDir "."))`.

To provide more hard data, here are these tests failing on TravisCI, on Linux and Max: https://travis-ci.com/github/wasp-lang/wasp/jobs/396611782#L536 , and you can see them in source here: https://github.com/wasp-lang/wasp/commit/86544c5917e63008c737d712fc25aeebd4ceb29c#diff-fc97c54b9046281d0b45c23a9382dca5R256-R265 .

and here are the ones failing on AppVeyor on Windows: https://ci.appveyor.com/project/Martinsos/wasp/builds/35627911#L3515 , with source being here: https://github.com/wasp-lang/wasp/commit/86544c5917e63008c737d712fc25aeebd4ceb29c#diff-fc97c54b9046281d0b45c23a9382dca5R267-R276 .

My big question would be: are Path.Windows and Path.Posix functions expected to behave differently according to the platform they are running on, or not? I understand that Path behaves differently regarding the platform it runs, and that is great because it enables us to work with system paths, but if Path.Windows and Path.Posix could guarantee consistent behaviour regardless of the platform, that would be great because it would allow us to work explicity with WIndows or Posix paths regardless of the platform we are running on (for example if we are parsing and crunching Posix paths from a log file but we are running on Windows, or vice versa). Right now, Path.Windows and Path.Posix don't seem to be completely platform independent.

**Comment by NorfairKing** [comment]
@Martinsos Ah I see what you mean now. Thanks for clarifying!

> are Path.Windows and Path.Posix functions expected to behave differently according to the platform they are running on, or not? 

That's exactly the right question. 

If it were up to me I'd say: YES, OBVIOUSLY.
I'm not sure how much responsibility I'm allowed to take here. I'm just one of the maintainers.

Let's ping @mrkkrp , @sjakobi and @harendra-kumar  and see what they have to say. I have some time this month to work on `path` and see if I can't get it cleaned up a bit.

**Comment by sjakobi** [comment]
> > are Path.Windows and Path.Posix functions expected to behave differently according to the platform they are running on, or not?
> 
> That's exactly the right question.
> 
> If it were up to me I'd say: YES, OBVIOUSLY.

@NorfairKing, Could you expand on why these modules should display platform-dependent behaviour? IMHO the entire point of having these modules was so paths could be manipulated independently of the current platform.

Here's a bit of insight regarding the different parsing of `.` on Linux:

```
> import Path.Internal
> let Path x = [PP.reldir|.|] in x
""
> let Path x = [PW.reldir|.|] in x
".\\"
```

Apparently `PW.normalizeDir` fails to properly normalize the "relative root path":

https://github.com/commercialhaskell/path/blob/a75ddfc86319cd2187eb8965bdb11b52690fa002/src/Path/Include.hs#L766-L776

I think the problem is that `relRootFP` is defined in a platform-dependent way in `Path.Internal`:

https://github.com/commercialhaskell/path/blob/a75ddfc86319cd2187eb8965bdb11b52690fa002/src/Path/Internal.hs#L60-L62

I think this can be fixed by moving `relRootFP` to `Path/Include.hs` and using the right `pathSeparator`.

**Comment by NorfairKing** [comment]
> @NorfairKing, Could you expand on why these modules should display platform-dependent behaviour? IMHO the entire point of having these modules was so paths could be manipulated independently of the current platform.

I misread the question. I think these modules should _NOT_ display platform-dependent behaviour.

**Comment by Martinsos** [comment]
@sjakobi great, sounds like that could be it!

Keep in mind there is one more possible unexpected behaviour, I described it in this comment above: https://github.com/commercialhaskell/path/issues/166#issuecomment-705237039 .

If you wish, I could try to find some time and write a PR for it (soonest this weekend), since you pretty clearly showed what the problem might be. Although I never contributed to this repo yet so it might take me some more time, but it is also a good way for me to get introduced to it, which I would like since I used Path a lot and am planning to use it on future projects also when needed.

**Comment by NorfairKing** [comment]
@Martinsos Sounds good, please ping me if you need any help.

**Comment by Martinsos** [comment]
@NorfairKing @sjakobi here is PR #167 !

**Comment by Martinsos** [comment]
Fixed with #167
